pkg/io/v2/opener.go

```diff
@@ -0,0 +1,363 @@
+	RangeReader(ctx context.Context, path string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error)
+	ReadObject(ctx context.Context, path string) ([]byte, error)
+	Attributes(ctx context.Context, path string) (attrs *blob.Attributes, err error)
+	SignedURL(ctx context.Context, path string, opts *blob.SignedURLOptions) (string, error)
+
+	ListSubPaths(ctx context.Context, path string, matchFns ...func(*blob.ListObject) bool) ([]string, error)
+	ListSubPathsIter(ctx context.Context, path string) (*blob.ListIterator, error)
+	ListSubDirs(ctx context.Context, path string) ([]string, error)
+
+	Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error)
+	Upload(ctx context.Context, uploads map[string]UploadFunc) error
+
+	// Workaround to retrieve the storageClient for prow/spyglass/testgrid.go because it calls
+	// github.com/GoogleCloudPlatform/testgrid/config.Read() which only works with GCS storage client right now
+	GetGCSClient(ctx context.Context) (*storage.Client, error)
+}
+
+type opener struct {
+	credentials []byte
+}
+
+// NewOpener returns an opener that can read GCS, S3 and local paths.
+func NewOpener(ctx context.Context, credentialsFile string) (Opener, error) {
+	var credentials []byte
+	var err error
+	if credentialsFile != "" {
+		credentials, err = ioutil.ReadFile(credentialsFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &opener{
+		credentials: credentials,
+	}, nil
+}
+
+func (o *opener) getBucket(ctx context.Context, path string) (*blob.Bucket, string, error) {
+	_, _, relativePath, err := providers.ParseStoragePath(path)
+	if err != nil {
+		return nil, "", fmt.Errorf("could not get bucket: %w", err)
+	}
+	bucket, err := providers.GetBucket(ctx, o.credentials, path)
+	if err != nil {
+		return nil, "", err
+	}
+	return bucket, relativePath, nil
+}
+
+func (o *opener) Reader(ctx context.Context, path string, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	reader, err := bucket.NewReader(ctx, relativePath, opts)
+	if err != nil {
+		return nil, err
+	}
+	return readerCloser{reader, bucket}, nil
+}
+
+type readerCloser struct {
+	r *blob.Reader
+	b *blob.Bucket
+}
+
+func (rc readerCloser) Read(p []byte) (n int, err error) {
+	return rc.r.Read(p)
+}
+
+func (rc readerCloser) Close() error {
+	if err := rc.r.Close(); err != nil {
+		return err
+	}
+	if err := rc.b.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *opener) RangeReader(ctx context.Context, path string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	reader, err := bucket.NewRangeReader(ctx, relativePath, offset, length, opts)
+	if err != nil {
+		return nil, err
+	}
+	return readerCloser{reader, bucket}, nil
```

Comment:
Why not just return the `reader`? It implements `io.ReadCloser`

Reply:
As discussed. I'll cache the bucket(s) and return the reader here